### PR TITLE
frame-session: Introduce a proper proof of key ownership

### PR DESCRIPTION
**Pull request by bkchr**
`frame_session::set_keys` supports providing a `proof` as second parameter. This `proof` was not yet checked. This pull request introduces a verification of this proof and also a way of generating this `proof`. The `proof` in FRAME are concatenated `Signatures`. These `Signatures` are in the same order as the public keys in the `SessionKeys` struct. Each signature is signing the `owner`, proofing that the generating party has access over the private key associated to the public session key. The `owner` in FRAME is the account id of the account that will call `set_keys`, aka the account of the `validator`.

This pull request is changing the `SessionKeys` runtime api. This still requires a RFC, as this is a public interface of the Polkadot runtime.

@jacogr polkadot-js should provide a way to use the runtime api directly to generate the sessions keys. 
@paritytech/docs-audit this will require updates of the validator documentation to tell them what to pass for `owner` and that they need to pass `proof` to `set_keys`.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3831963

**Comment by bkchr** [comment]
bot fmt

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3833890

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3834074

File: substrate/client/rpc/src/author/mod.rs [hunk 1]
```diff
@@ -123,10 +123,25 @@ where
...
+		let res = if version < 2 {
+			#[allow(deprecated)]
+			runtime_api
+				.generate_session_keys_before_version_2(best_block_hash, None)
+				.map(Into::into)
+				.map_err(|api_err| Error::Client(Box::new(api_err)))
+		} else {
+			runtime_api
+				.generate_session_keys(best_block_hash, Vec::new(), None)
+				.map(|sk| sk.keys.into())
```
Comment (davxy) [inline substrate/client/rpc/src/author/mod.rs:140]: If I understood correctly, we may want to change this RPC call to return an encoded `GeneratedSessionKeys` right? (Probably with an empty proofs vec if version < 2).

That is, also return the proofs so that the user can eventually submit the `set-keys` extrinsic to the session pallet which now requires the proofs as well.

Otherwise the user should manually (re)generate these proofs before on-chain keys registration

File: substrate/frame/staking/src/tests.rs [hunk 2]
```diff
@@ -5308,6 +5317,7 @@ fn min_commission_works() {
 
 #[test]
 #[should_panic]
+#[cfg(debug_assertions)]
```
[the remark this answers is not in the document]
Comment (bkchr) [reply substrate/frame/staking/src/tests.rs:5320]: Because they are failing if not running with debug asserts enabled. Aka if you just run `cargo test --all --release`. In CI we also enable debug asserts in release mode.

File: substrate/client/rpc/src/author/mod.rs (hunk 1 above)
Comment (bkchr) [reply substrate/client/rpc/src/author/mod.rs:140]: > If I understood correctly, we may want to change this RPC call to return an encoded `GeneratedSessionKeys` right?

No. The RPC should not be used anymore. People should just directly call the runtime api using `state_call`.

Comment (davxy) [reply substrate/client/rpc/src/author/mod.rs:140]: Clear

**Comment by bkchr** [comment]
> @jacogr polkadot-js should provide a way to use the runtime api directly to generate the sessions keys.

Scrape that, we need a new RPC for this.

**Comment by kianenigma** [comment]
> https://github.com/orgs/paritytech/teams/docs-audit this will require updates of the validator documentation to tell them what to pass for owner and that they need to pass proof to set_keys.

This is mostly in the Polkadot wiki, I will ping the w3f folks and also setup a system where they can listen to mentioned of this `docs-audit` team. But all in all, you should do your best to update the docs around this in the rust docs of pallet session at first.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3870331

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3870483

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3929724

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3939449

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3939719

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3943055

File: substrate/client/rpc-api/src/author/mod.rs [hunk 3]
```diff
@@ -41,6 +55,15 @@ pub trait AuthorApi<Hash, BlockHash> {
 	#[method(name = "author_rotateKeys")]
 	fn rotate_keys(&self) -> Result<Bytes, Error>;
 
+	/// Generate new session keys and returns the corresponding public keys.
+	///
+	/// The `owner` should be something that can be used on chain for verifying the ownership of the
+	/// generated keys using the returned `proof`. For example `owner` could be set to the account
+	/// id of the account registering the returned public session keys. The actual data to pass for
+	/// `owner` depends on the runtime logic verifying the `proof`.
```
Comment (davxy) [inline substrate/client/rpc-api/src/author/mod.rs:63]: Maybe we can add an example here?

Something like: "For example, "session pallet" `set_keys` requires `owner` to be the account used to sign the extrinsic."

File: cumulus/parachains/runtimes/starters/seedling/src/lib.rs [hunk 4]
```diff
@@ -354,8 +354,11 @@ impl_runtime_apis! {
 	}
 
 	impl sp_session::SessionKeys<Block> for Runtime {
-		fn generate_session_keys(seed: Option<Vec<u8>>) -> Vec<u8> {
-			SessionKeys::generate(seed)
+		fn generate_session_keys(_: Vec<u8>, _: Option<Vec<u8>>) -> sp_session::OpaqueGeneratedSessionKeys {
+			sp_session::OpaqueGeneratedSessionKeys {
```
Comment (davxy) [inline cumulus/parachains/runtimes/starters/seedling/src/lib.rs:358]: Nitpick. Makes sense to implement `Default` for `OpaqueGeneratedSessionKeys`? I see a couple of use cases around

**Review by davxy — APPROVED** [review]
Looks good

**Comment by ordian** [comment]
cc @drskalman @alistairstewart

**Comment by drskalman** [comment]
This is a very important work and I'm not sure if my comment should be considered after its merger for future improvement or not.

1. Proof of possession with BLS signature is secure only if it is done with different signature scheme than the one is used by the signer to sign other messages (using different hash function for example). The attack is described on page 15 of [1].

- [1] Ristenpart, T., & Yilek, S. (2007). The power of
  proofs-of-possession: Securing multiparty signatures against
  rogue-key attacks. In , Annual {{International Conference}} on the
  {{Theory}} and {{Applications}} of {{Cryptographic Techniques} (pp.
  228–245). : Springer.

2. In our flavor of BLS signature, The verification of proof of possession function not only the verifies that the signature is correct but also verifies that the signer has used the same private key to generate the public keys in $G1$ and $G2$: 
https://github.com/w3f/bls/blob/master/src/double_pop.rs#L67. This is not being tested by calling normal verify on the signature.

In that regard, it seems that the correct way of implementing the macro is perhaps to implement the proof of possession generation and verification inside each crypto module and then call that function instead of generic sign and verify and perhaps call generic sign and verify inside that function for all key types except for BLS.

3. @AlistairStewart is of opinion that everybody in the world calls this "Proof of Possession" and not "Ownership proof". Calling it "ownership proof" make it very hard for any outsider to spot and find the code. I'm not sure what is cost of renaming these functions though.

File: substrate/primitives/runtime/src/traits.rs [hunk 5]
```diff
@@ -2045,10 +2073,18 @@ macro_rules! impl_opaque_keys_inner {
 						>::generate_pair(seed.clone()),
 					)*
 				};
-				$crate::codec::Encode::encode(&keys)
+
+				let proof = keys.create_ownership_proof(owner)
```
Comment (coax1d) [inline substrate/primitives/runtime/src/traits.rs:2205]: @drskalman is this `create_ownership_proof` function the one that we just replace with our new pop implementation. It is called a few times within this PR in various places.

**Comment by seunlanlege** [comment]
WEN

**Comment by bkchr** [comment]
@seunlanlege please stop spamming. This issue is not a requirement for BLS.

**Comment by bkchr** [comment]
https://github.com/paritytech/polkadot-sdk/pull/9471 this was the last pr from @drskalman. This pr brings a feature that just improves the security of `set_keys`.

**Comment by Ank4n** [comment]
> `frame_session::set_keys` supports providing a `proof` as second parameter. This `proof` was not yet checked. This pull request introduces a verification of this proof and also a way of generating this `proof`. The `proof` in FRAME are concatenated `Signatures`. These `Signatures` are in the same order as the public keys in the `SessionKeys` struct. Each signature is signing the `owner`, proofing that the generating party has access over the private key associated to the public session key. The `owner` in FRAME is the account id of the account that will call `set_keys`, aka the account of the `validator`.

@bkchr will this break staking proxy accounts (including pure proxied accounts) from being able to call `set_keys`?

In other words, does the session key need to be generated with the validator stash as the owner, and does that mean only the validator stash itself can make the `set_keys` call?  

Am I understanding this correctly, or am I missing something?

**Comment by Ank4n** [comment]
> @bkchr will this break staking proxy accounts (including pure proxied accounts) from being able to call `set_keys`?
> 
> In other words, does the session key need to be generated with the validator stash as the owner, and does that mean only the validator stash itself can make the `set_keys` call?
> 
> Am I understanding this correctly, or am I missing something?

Oh wait I think it will work. Am dumb. 
The proof is signed by the session keys, which include the owner public key, and a proxied `set_keys` call would be re-dispatched with the same owner (the validator stash).
Still, I’ll wait for your confirmation to be sure.

**Comment by bkchr** [comment]
> The proof is signed by the session keys, which include the owner public key, and a proxied `set_keys` call would be re-dispatched with the same owner (the validator stash).

Yes exactly, the session keys are signing some "owner". In our case we are signing the bytes of the accountid being the origin for `set_keys`. So, you would use the `stash_address`. It is just to proof that the guy calling `set_keys` actually has access to the private session keys.

TLDR: Nothing should break.
